src/condition/cmdi_detector.cpp

```diff
@@ -0,0 +1,263 @@
+};
+
+// Most shells support -c as a way to specify a shell command, however some
+// shells such as ksh allow for the first argument to be a shell command
+std::unordered_map<std::string_view, std::string_view> known_shells{
+    {"sh", "c"},
+    {"bash", "c"},
+    {"ksh", {}},
+    {"rksh", {}},
+    {"fish", "c"},
+    {"zsh", "c"},
+    {"dash", "c"},
+    {"ash", "c"},
+    {"powershell", "Command"},
+    {"powershell.exe", "Command"},
+};
+
+std::string_view basename(std::string_view path)
+{
+    auto idx = path.find_last_of(R"(\/)"sv);
+    return idx == std::string_view::npos ? path : path.substr(idx + 1);
+}
+
+std::string_view trim_whitespaces(std::string_view str)
+{
+    static const std::string_view whitespaces = " \f\n\r\t\v";
+
+    if (str.empty()) {
+        return str;
+    }
+
+    auto start = str.find_first_not_of(whitespaces);
+    if (start == std::string_view::npos) {
+        return str;
+    }
+
+    auto end = str.find_last_not_of(whitespaces);
+    return str.substr(start, 1 + end - start);
+}
+
+std::size_t object_size(const ddwaf_object &obj) { return static_cast<std::size_t>(obj.nbEntries); }
+
+std::string_view object_at(const ddwaf_object &obj, std::size_t idx)
+{
+    const ddwaf_object &child = obj.array[idx];
+    if (child.type != DDWAF_OBJ_STRING) {
+        return {};
+    }
+    return {child.stringValue, object_size(child)};
+}
+
+std::string_view find_shell_command(std::string_view executable, const ddwaf_object &exec_args)
+{
+    // By initialising the cow_string with the underlying data
+    // contained within the std::string, we ensure a new one won't be
+    // allocated once the string is modified.
+    cow_string executable_lc{executable};
+    if (transformer::lowercase::transform(executable_lc)) {
+        executable = static_cast<std::string_view>(executable_lc);
+    }
+
+    auto shell_it = known_shells.find(basename(executable));
+    if (shell_it != known_shells.end()) {
+        // We've found that the current exec command is attempting to run a
+        // a shell. The shell binary itself might be injected, but also the
+        // shell command. So we need to identify the command
+        std::size_t i = 1;
+        if (!shell_it->second.empty()) {
+            // Most shells allow specifying a command with -c
+            for (; i < object_size(exec_args); ++i) {
+                auto opt = trim_whitespaces(object_at(exec_args, i));
+                if (!opt.empty() && opt[0] == '-' &&
+                    opt.find(shell_it->second) != std::string_view::npos) {
+                    // We've found the -c option, we can now break, if it isn't found
+                    // i will reach the end of the array
+                    break;
+                }
+            }
+        }
+        for (; i < object_size(exec_args); ++i) {
+            auto arg = trim_whitespaces(object_at(exec_args, i));
+            if (arg.empty() || arg.starts_with('-') || arg.starts_with('+')) {
```

Comment:
what's with the plus? what shells support this? You're missing valid commands in powershell, at least:

```
$ pwsh -Command '+1; Write-Output "Hello" '
1
Hello
```

Reply:
Not sure about powershell (I guess not), but you can use `shopts` in the terminal with most linux shells, e.g. `bash -e -c 'false ; ls -l'` and the equivalent with errexit disabled `bash +e -c 'false ; ls -l'`. I have a few patterns which aren't currently working so I'll have to update a bit the way we interpret command line arguments.